File: apps/ios/Sources/ParkShieldFoundation/Core/Network/APIError.swift
import Foundation

struct APIError: Decodable, Equatable, Sendable {
    let version: String
    let code: String
    let message: String
    let correlationID: String
    let details: [APIErrorDetail]?

    enum CodingKeys: String, CodingKey {
        case version, code, message, details
        case correlationID = "correlation_id"
    }
}

struct APIErrorDetail: Decodable, Equatable, Sendable {
    let field: String
    let code: String
}

enum APIErrorContract {
    static let version = "1"
    private static let allowedDetailCodes: Set<String> = ["MISSING_FIELD", "INVALID_FIELD"]
    private static let correlationIDPattern = "^[A-Za-z0-9._:-]{1,128}$"
    private static let fieldPattern = "^[A-Za-z0-9_.\\[\\]-]{1,128}$"

    static func isValid(_ error: APIError) -> Bool {
        guard error.version == version,
              !error.code.isEmpty,
              !error.message.isEmpty,
              error.correlationID.range(of: correlationIDPattern, options: .regularExpression) != nil
        else { return false }

        return (error.details ?? []).allSatisfy {
            $0.field.range(of: fieldPattern, options: .regularExpression) != nil &&
                allowedDetailCodes.contains($0.code)
        }
    }
}

File: apps/android/app/src/main/kotlin/ai/parkshield/android/core/network/ApiError.kt
package ai.parkshield.android.core.network

data class ApiError(
    val version: String,
    val code: String,
    val message: String,
    val correlationId: String,
    val details: List<ApiErrorDetail> = emptyList(),
)

data class ApiErrorDetail(val field: String, val code: String)

object ApiErrorContract {
    const val VERSION = "1"
    private val allowedDetailCodes = setOf("MISSING_FIELD", "INVALID_FIELD")

    fun isValid(error: ApiError): Boolean =
        error.version == VERSION &&
            error.code.isNotBlank() &&
            error.message.isNotBlank() &&
            error.correlationId.matches(Regex("[A-Za-z0-9._:-]{1,128}")) &&
            error.details.all { it.field.matches(Regex("[A-Za-z0-9_.\\[\\]-]{1,128}")) && it.code in allowedDetailCodes }
}
